src/zope/testrunner/find.py

```diff
@@ -418,11 +442,19 @@ def tests_from_suite(suite, options, dlevel=1,
     if isinstance(suite, unittest.TestSuite):
         for possible_suite in suite:
             for r in tests_from_suite(possible_suite, options, level, layer,
-                                      accept=accept):
+                                      accept=accept,
+                                      seen_test_ids=seen_test_ids,
+                                      duplicated_test_ids=duplicated_test_ids):
                 yield r
     elif isinstance(suite, StartUpFailure):
         yield (suite, None)
     else:
+        if options.require_unique_ids:
+            suite_id = suite.id()
+            if suite_id in seen_test_ids:
+                duplicated_test_ids.add(suite_id)
+            else:
+                seen_test_ids.add(suite_id)
         if level <= options.at_level:
             if accept is None or accept(str(suite)):
                 yield (suite, layer)
```

Comment:
Note that you're enforcing unique `suite.id()`, but the filtering checks `str(suite)`.  These are different:

- for regular unittest.TestCase tests `id()` returns `package.module.ClassName.test_method`, while `str()` returns `test_method (package.module.ClassName)`
- for doctest.DocFileSuite tests, `id()` returns `filename_txt`, while `str()` returns `/absolute/path/to/the/filename.txt`.

(Incidenally, this is why the original bug report doesn't apply to the test runner: we can always distinguish different distribution.txt doctest files if they're in different directories).